lib/rspec/core/reporter.rb

```diff
@@ -79,6 +88,22 @@ def message(message)
       notify :message, Notifications::MessageNotification.new(message)
     end
 
+    # @param event [Symbol] Name of the custom event to trigger on formatters
+    # @param options [Hash] Hash of arguments to provide via `CustomNotification`
+    #
+    # Publish a custom event to supporting registered formatters.
+    # @see RSpec::Core::Notifications::CustomNotification
+    def publish(event, options={})
+      raise UnsupportedPublish if RSPEC_NOTIFICATIONS.include? event
+      notify event, Notifications::CustomNotification.for(options)
+    end
+
+    # @private
+    UnsupportedPublish = StandardError.new(
+      "RSpec::Core::Reporter#publish is intended for sending custom " \
+      "events not internal RSpec ones, please rename your custom event."
+    )
```

Comment:
I can go back to a bare raise with a string if required, the thinking behind the singleton was just to save allocations.